src/class/dfu/dfu_device.c

```diff
@@ -0,0 +1,648 @@
+  {
+    case DFU_REQUEST_DNLOAD:
+    {
+      if ( (stage == CONTROL_STAGE_ACK)
+           && ((_dfu_state_ctx.attrs & DFU_FUNC_ATTR_CAN_DOWNLOAD_BITMASK) != 0)
+           && (_dfu_state_ctx.state == DFU_DNLOAD_SYNC))
+      {
+        dfu_req_dnload_reply(rhport, request);
+        return true;
+      }
+    } // fallthrough
+    case DFU_REQUEST_DETACH:
+    case DFU_REQUEST_UPLOAD:
+    case DFU_REQUEST_GETSTATUS:
+    case DFU_REQUEST_CLRSTATUS:
+    case DFU_REQUEST_GETSTATE:
+    case DFU_REQUEST_ABORT:
+    {
+      if(stage == CONTROL_STAGE_SETUP)
+      {
+        return dfu_state_machine(rhport, request);
+      }
+    }
+    break;
+
+    default:
+    {
+      TU_LOG2("  DFU Nonstandard Request: %u\r\n", request->bRequest);
+      return false; // stall unsupported request
+    }
+    break;
+  }
+
+  return true;
+}
+
+static uint16_t dfu_req_upload(uint8_t rhport, tusb_control_request_t const * request, uint16_t block_num, uint16_t wLength)
+{
+  TU_VERIFY( wLength <= CFG_TUD_DFU_TRANSFER_BUFFER_SIZE);
+  uint16_t retval = tud_dfu_req_upload_data_cb(block_num, (uint8_t *)&_dfu_state_ctx.transfer_buf, wLength);
+  tud_control_xfer(rhport, request, &_dfu_state_ctx.transfer_buf, retval);
+  return retval;
+}
+
+static void dfu_req_getstatus_reply(uint8_t rhport, tusb_control_request_t const * request)
+{
+  dfu_status_req_payload_t resp;
+
+  resp.bStatus = _dfu_state_ctx.status;
+  memset((uint8_t *)&resp.bwPollTimeout, 0x00, 3);
+  resp.bState = _dfu_state_ctx.state;
+  resp.iString = 0;
+
+  tud_control_xfer(rhport, request, &resp, sizeof(dfu_status_req_payload_t));
+}
+
+static void dfu_req_getstate_reply(uint8_t rhport, tusb_control_request_t const * request)
+{
+  tud_control_xfer(rhport, request, &_dfu_state_ctx.state, 1);
+}
+
+static void dfu_req_dnload_setup(uint8_t rhport, tusb_control_request_t const * request)
+{
+  // TODO: add "zero" copy mode so the buffer we read into can be provided by the user
+  // if they wish, there still will be the internal control buffer copy to this buffer
+  // but this mode would provide zero copy from the class driver to the application
+
+  // setup for data phase
+  tud_control_xfer(rhport, request, &_dfu_state_ctx.transfer_buf, request->wLength);
+}
+
+static void dfu_req_dnload_reply(uint8_t rhport, tusb_control_request_t const * request)
+{
+  (void) rhport;
+  tud_dfu_req_dnload_data_cb(request->wValue, (uint8_t *)&_dfu_state_ctx.transfer_buf, request->wLength);
+  _dfu_state_ctx.blk_transfer_in_proc = false;
+}
+
+void tud_dfu_dnload_complete(void)
+{
+  if (_dfu_state_ctx.state == DFU_DNBUSY)
+  {
+    _dfu_state_ctx.state = DFU_DNLOAD_SYNC;
+  } else if (_dfu_state_ctx.state == DFU_MANIFEST)
+  {
+    _dfu_state_ctx.state = ((_dfu_state_ctx.attrs & DFU_FUNC_ATTR_MANIFESTATION_TOLERANT_BITMASK) != 0)
+                           ? DFU_MANIFEST_WAIT_RESET : DFU_MANIFEST_SYNC;
+  }
+}
+
+static bool dfu_state_machine(uint8_t rhport, tusb_control_request_t const * request)
```

Comment:
IMHO, dfu_state_machine() is written more complicated than it should be. To be honest, I don't know what would suggest here. Maybe when adding example we could understand this better.

Reply:
Yes, so this concern one of my prior comments.  I am very open to changing it, but here is some background for how it came to be.

I opted for the current setup where there is a very verbose state machine, but it is explicit in the behaviour to expect on any given request.  An earlier iteration had all behaviour inside of dfu_moded_control_xfer_cb.  This was incredibly difficult to follow, since each request ended up containing a switch for every single possible state, where most resulted in a stall.

I then migrated to a setup where some common commands such as DFU_REQUEST_GETSTATUS and DFU_REQUEST_GETSTATE where handled in dfu_moded_control_xfer_cb in the bRequest switch, and the rest where in the dfu_state_machine function.  This was better, but there still was some ugliness due to DFU_MANIFEST, DFU_MANIFEST_WAIT_RESET, and DFU_DNBUSY supposed to stall when these requests occur or DFU_MANIFEST_SYNC having different behaviour on a DFU_REQUEST_GETSTATUS if DFU_FUNC_ATTR_MANIFESTATION_TOLERANT_BITMASK is set.

Reply:
I see, it is your point of view for implementation. Though it is still a bit complicated for my taste. However, you don't have to change it now. We will see how it turn out when adding the example. And I will try to move code around if needed as follow up later on. I do have some renaming in my mind. It is too minor mentioned in reviews.